tests/src/test/scala/scalaz/DisjunctionTest.scala

```diff
@@ -32,4 +32,41 @@ object DisjunctionTest extends SpecLite {
     \/.fromTryCatchThrowable[Int, Foo](throw bar) must_=== \/.left(bar)
     \/.fromTryCatchThrowable[Int, Bar](throw foo).mustThrowA[Foo]
   }
+
+  "recover" in {
+    sealed trait Foo
+    final class Bar extends Foo
+    final class Baz extends Foo
+    val bar = new Bar
+
+    implicit val equalFoo = Equal.equalA[Foo]
+    implicit val showFoo = Show.showA[Foo]
+
+    -\/[Foo](bar).recover({ case _: Bar => 1 }) must_=== \/-(1)
```

Comment:
even though this is just a test, making Bar and Baz case objects would mean we don't have ugly type casing.

I'd hope we can eliminate type casing entirely, perhaps using wart-remover or the like.

    sealed trait Foo
    case object Bar extends Foo
    case object Baz extends Foo

    -\/[Foo](Bar).recover({ case Bar => 1 }) must_=== \/-(1)